### PR TITLE
Parallel catalog traversal in garbage collection

### PR DESCRIPTION
**Pull request by jblomer**
Replaces #2471

### CODE REVIEW

**Comment by jblomer** [comment]
@cernvm-bot cloudtest full

**Comment by janpriessnitz** [comment]
@cernvm-bot unittest full

**Comment by janpriessnitz** [comment]
@cernvm-bot unittest full

**Comment by janpriessnitz** [comment]
@cernvm-bot cloudtest cc7 nodestroy

**Comment by janpriessnitz** [comment]
@cernvm-bot cloudtest cc7 nodestroy

**Comment by janpriessnitz** [comment]
@cernvm-bot cloudtest cc7 nodestroy

**Comment by janpriessnitz** [comment]
@cernvm-bot cloudtest full nodestroy

File: cvmfs/garbage_collection/garbage_collector.h [hunk 1]
```diff
@@ -122,16 +122,22 @@ class GarbageCollector {
...
-    virtual ~ReflogBasedInfoShim() { }
+    explicit ReflogBasedInfoShim(ReflogTN *reflog) : reflog_(reflog) {
+      pthread_mutex_init(&reflog_mutex_, NULL);
+    }
+    virtual ~ReflogBasedInfoShim() {
+      pthread_mutex_destroy(&reflog_mutex_);
+    }
     virtual uint64_t GetLastModified(const CatalogTN *catalog) {
       uint64_t timestamp;
+      MutexLockGuard m(&reflog_mutex_);
```
Comment (jblomer) [inline cvmfs/garbage_collection/garbage_collector.h:133]: Excellent catch!

**Comment by jblomer** [comment]
@cernvm-bot cloudtest

**Comment by jblomer** [comment]
@cernvm-bot cloudtest full

**Comment by jblomer** [comment]
@cernvm-bot cloudtest full

**Pull request merged**